misc/testerina/modules/testerina-core/src/main/java/org/ballerinalang/testerina/natives/test/FunctionMock.java

```diff
@@ -0,0 +1,225 @@
+                mockFuncObj.getStringValue("functionToMock");
+        String originalFunctionPackage =
+                mockFuncObj.getStringValue("functionToMockPackage");
+        originalFunctionPackage = formatFunctionPackage(originalFunctionPackage);
+        String version =
+                mockFuncObj.getStringValue("packageVersion");
+        Object returnVal = null;
+
+        for (String caseId : caseIds) {
+            if (MockRegistry.getInstance().hasCase(caseId)) {
+                returnVal = MockRegistry.getInstance().getReturnValue(caseId);
+                if (returnVal.toString().contains(MockConstants.FUNCTION_CALL_PLACEHOLDER)) {
+                    return callFunction(originalFunction, originalFunctionPackage, returnVal.toString(), version, args);
+                }
+                break;
+            }
+        }
+        if (returnVal == null) {
+            String detail = "no return value or action registered for function";
+            return BallerinaErrors.createError(MockConstants.FUNCTION_CALL_ERROR, detail);
+        }
+        return returnVal;
+    }
+
+    private static Object callFunction(String originalFunction, String originalFunctionPackage, String returnVal,
+                                       String version, Object... args) {
+        int prefixPos = returnVal.indexOf(MockConstants.FUNCTION_CALL_PLACEHOLDER);
+        String methodName = returnVal.substring(prefixPos + MockConstants.FUNCTION_CALL_PLACEHOLDER.length());
+        Strand strand = Scheduler.getStrand();
+
+        String className;
+        String orgName;
+        String packageName;
+
+        // Set project info
+        try {
+            String[] projectInfo =
+                    Thread.currentThread().getStackTrace()[4].getClassName().split(Pattern.quote("."));
+            orgName = projectInfo[0];
+            packageName = projectInfo[1];
+            className = "tests." +
+                    getClassName(methodName, orgName, packageName, version, originalFunction, originalFunctionPackage);
+        } catch (IOException | ClassNotFoundException e) {
+            return BallerinaErrors.createError(MockConstants.FUNCTION_CALL_ERROR, e.getMessage());
+        }
+
+        List<Object> argsList = new ArrayList<>();
+        for (Object arg : args) {
+            argsList.add(arg);
+        }
+
+        ClassLoader classLoader = FunctionMock.class.getClassLoader();
+        return Executor.executeFunction(strand.scheduler, classLoader, orgName,
+                                        packageName, className, methodName, argsList.toArray());
+    }
+
+    private static String getClassName(String mockMethodName, String orgName, String packageName, String version,
+                                       String originalMethodName, String originalPackageName)
+            throws IOException, ClassNotFoundException {
+        String jarName = orgName + "-" + packageName + "-" + version + "-testable.jar";
```

Comment:
Can't we utilize the classloader for getting the classname?

Reply:
I worked on implementing this using the class loader and it worked fine for Mock functions defined in the same file. However, when moving mock functions to a different file, the implementation fails. 
Eventually, I settled for directly obtaining the version and passing it from the compile-time desugar phase to the runtime and it works. 

This might have to be revisited